include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
+	}
+};
+template<typename Vector>
+requires concepts::FloatingPointVector<Vector>
+struct length_helper<Vector>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	static inline typename vector_traits<Vector>::scalar_type __call(NBL_CONST_REF_ARG(Vector) vec)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findUMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return std::sqrt(dot_helper<Vector>::__call(vec, vec));
 	}
 };
-
-template<int N>
-struct find_msb_helper<vector<int32_t, N> >
+template<typename Vectorial>
+requires concepts::FloatingPointLikeVectorial<Vectorial>
+struct normalize_helper<Vectorial>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	static inline Vectorial __call(NBL_CONST_REF_ARG(Vectorial) vec)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findSMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return vec / length_helper<Vectorial>::__call(vec);
 	}
 };
 
-#ifndef __HLSL_VERSION
-
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct find_lsb_helper<T>
+{
+	using return_t = int32_t;
+	static inline T __call(const T arg)
+	{
+		return glm::findLSB<T>(arg);
+	}
+};
+template<typename Integer>
+NBL_PARTIAL_REQ_TOP(concepts::IntegralScalar<Integer>)
+struct find_msb_helper<Integer NBL_PARTIAL_REQ_BOT(concepts::IntegralScalar<Integer>) >
+{
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(Integer) val)
+	{
+		return glm::findMSB<Integer>(val);
+	}
+};
+// TODO: implemet to be compatible with both C++ and HLSL when it works with DXC
+template<typename EnumType>
+requires std::is_enum_v<EnumType>
+struct find_lsb_helper<EnumType>
+{
+	using return_t = int32_t;
+	static int32_t __call(NBL_CONST_REF_ARG(EnumType) val)
+	{
+		using underlying_t = std::underlying_type_t<EnumType>;
+		return find_lsb_helper<underlying_t>::__call(static_cast<underlying_t>(val));
+	}
+};
 template<typename EnumType>
-	requires std::is_enum_v<EnumType>
+requires std::is_enum_v<EnumType>
 struct find_msb_helper<EnumType>
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(EnumType) val)
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(EnumType) val)
 	{
 		using underlying_t = std::underlying_type_t<EnumType>;
-		return find_msb_helper<underlying_t>::findMSB(static_cast<underlying_t>(val));
+		return find_msb_helper<underlying_t>::__call(static_cast<underlying_t>(val));
 	}
 };
 
-#endif
+template<typename FloatingPoint>
+requires concepts::FloatingPointScalar<FloatingPoint>
+struct rsqrt_helper<FloatingPoint>
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) x)
+	{
+		// TODO: https://stackoverflow.com/a/62239778
+		return 1.0f / std::sqrt(x);
+	}
+};
 
-template<typename Integer>
-struct find_lsb_helper;
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct frac_helper<T>
+{
+	using return_t = T;
+	static inline return_t __call(const T x)
+	{
+		return x - std::floor(x);
```

Comment:
I think `std::modf` and an if statement to make sure negatives behave as they should (1-modf fraction) is more accurate than x-floor(x)

Reply:
according to spir-v spec its x - floor(x)
![image](https://github.com/user-attachments/assets/ce76ceab-d432-43c9-9550-a1c03a712412)

Reply:
ok then lets follow SPIR-V